pkg/apis/flows/v1beta1/sequence_types.go

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2020 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	messagingv1beta1 "knative.dev/eventing/pkg/apis/messaging/v1beta1"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// Sequence defines a sequence of Subscribers that will be wired in
+// series through Channels and Subscriptions.
+type Sequence struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Sequence.
+	Spec SequenceSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Sequence. This data may be out of
+	// date.
+	// +optional
+	Status SequenceStatus `json:"status,omitempty"`
+}
+
+var (
+	// Check that Sequence can be validated and defaulted.
+	_ apis.Validatable = (*Sequence)(nil)
+	_ apis.Defaultable = (*Sequence)(nil)
+
+	// Check that Sequence can return its spec untyped.
+	_ apis.HasSpec = (*Sequence)(nil)
+
+	// TODO: make appropriate fields immutable.
+	//_ apis.Immutable = (*Sequence)(nil)
+
+	_ runtime.Object = (*Sequence)(nil)
+
+	// Check that we can create OwnerReferences to a Sequence.
+	_ kmeta.OwnerRefable = (*Sequence)(nil)
+)
+
+type SequenceSpec struct {
+	// Steps is the list of Destinations (processors / functions) that will be called in the order
+	// provided.
+	Steps []duckv1.Destination `json:"steps"`
```

Comment:
to discuss: https://github.com/knative/eventing/issues/2302. My main argument is one of the reasons for retrying is because a service depends on a backend service that is known to be not very reliable. And one reason to limit retrying is because it might cost real money each time a call is made. That's why I'm arguing for fine-grained delivery control.